### DISCUSSION TITLE
Feature proposal: Layer shuffling

### DISCUSSION THREAD

**Discussion by foldl** [ggml-org/llama.cpp#7155] [Ideas]
Llama-3 70B has been self-merged by @mlabonne into a [120B](https://huggingface.co/mlabonne/Meta-Llama-3-120B-Instruct) model. Someone reported that it performs quite well. He even [apologized](https://twitter.com/maximelabonne/status/1787056002640028015) in advance if AGI is achieved by duplicating random layers. On the other hand, others had tried to delete some layers, such as [here](https://twitter.com/kwindla/status/1788224280754618393).

Layer duplicating or removal can be done on-the-fly. So I do it in chatllm.cpp:

https://github.com/foldl/chatllm.cpp/commit/fb8690cebb8ce9e1548db438930d0da48624abbc

Doc: https://github.com/foldl/chatllm.cpp/blob/master/docs/fun.md#layer-shuffling

I think it would be nice for llama.cpp to have this functionality, too. Why? It is fun, isn't it?

**Comment by marcingomulkiewicz** [comment]
I believe you mean something similar to what I have suggested in #7012 . :slightly_smiling_face:

**Reply by foldl** [reply]
Yes, exactly.

Maybe this thread can be kept open for a while since an implementation is provided.
